docs/solutions/rhel74_image/ipxe.sh.template

```diff
@@ -0,0 +1,161 @@
+service firewalld restart
+
+systemctl enable httpd
+systemctl start httpd
+
+mkdir /var/www/isos
+mkdir /mnt/iso
+mkdir /var/www/html/ipxe
+mkdir /var/www/html/oci
+mkdir /var/www/html/<OS_NAME>
+mkdir -p /home/opc/.oci
+
+chmod 755 /home/opc/.oci
+chmod 777 /var/www/isos
+
+easy_install pip
+pip install pip --upgrade
+pip install oci-cli
+
+sed '/imtcp/s/#//g' -i.bak /etc/rsyslog.conf
+sed '/InputTCPServerRun/s/#//g' -i.bak /etc/rsyslog.conf
+systemctl restart rsyslog.service
+
+cd /home/opc/.oci
+if [ -z "${PR_KEY_PW}" ]
+then
+	cat > config <<-EOF
+[DEFAULT]
+user=${USER}
+fingerprint=${FINGERPRINT}
+key_file=/home/opc/.oci/oci_api_key.pem
+tenancy=${TENANCY}
+region=${REGION}
+EOF
+else
+	cat > config <<-EOF
+[DEFAULT]
+user=${USER}
+fingerprint=${FINGERPRINT}
+key_file=/home/opc/.oci/oci_api_key.pem
+tenancy=${TENANCY}
+region=${REGION}
+pass_phrase=${PR_KEY_PW}
+EOF
+fi
+
+privkey
+
+chown -R opc:opc /home/opc/.oci
+
+cd /var/www/html/ipxe
+ks
+sed -i.bak 's|<IP>|'"${LOCAL_IP}"'|g
+s|<OS>|'"${OS}"'|g
+s|<RHEL_USER>|'"${RU}"'|g
+s|<RHEL_PW>|'"${RP}"'|g' ks.cfg
+
+cat > ipxeboot <<-EOF
+#!ipxe
+set target-iqn iqn.2015-02.oracle.boot:uefi
+set root-path iscsi:169.254.0.2::::${target-iqn}
+set alt-root iscsi:169.254.0.2:::1:${target-iqn}
+sanhook ${root-path} || sanhook ${alt-root}
+set uefi-filename \EFI\BOOT\grubx64.efi 
+sanboot || set url http://${LOCAL_IP}
+# Set your kernel and initrd here...
+kernel \${url}/<OS_NAME>/images/pxeboot/vmlinuz initrd=initrd.img vmlinuxconsole=tty0 console=ttyS0,9600 netroot=iscsi:@169.254.0.2::::iqn.2015-02.oracle.boot:uefi iscsi_initiator=iqn.2015-02.oracle.boot:instance ip=dhcp ks=http://${LOCAL_IP}/ipxe/ks.cfg
+initrd \${url}/<OS_NAME>/images/pxeboot/initrd.img
+boot
+EOF
+
+cd /var/www/html/oci
+cloud
+firewallcfg
+
+cat > /home/opc/sacrifice-launch.sh.template <<-EOF
+#!/bin/bash
+sac_inst_ocid=\`oci compute instance launch --availability-domain=${AD} \
+--compartment-id=${COMPARTMENT} --image-id=${SAC_OCID} \
+--display-name="sacrifice" --ipxe-script-file="/var/www/html/ipxe/ipxeboot" \
+--shape=${SHAPE} --subnet-id=<SUBNET_OCID> \
+--metadata='{"ssh_authorized_keys": "${SSH_PUBLIC_KEY}" }' | jq -r '.data["id"]'\`
+sleep 120
+sac_vnic_attach=\`oci compute vnic-attachment list --instance-id=\${sac_inst_ocid} \
+--compartment-id=${COMPARTMENT} | jq -r '.data[0]["vnic-id"]'\`
+sac_ip=\`oci network vnic get --vnic-id=\${sac_vnic_attach} | jq -r '.data["private-ip"]'\`
+build_ping=1
+while [ \${build_ping} == 1 ]
+do
+	sleep 10
+	ping -q -c 1 \${sac_ip}
+	build_ping=\$?
+done
+	
+ping_result=0
+while [ \${ping_result} == 0 ]
+do
+	sleep 10
+	ping -q -c 1 \${sac_ip}
+	ping_result=\$?
+done
+image_ocid=\`oci compute image create --compartment-id=${COMPARTMENT} --instance-id=\${sac_inst_ocid} \
+--display-name="RHEL_74" | jq -r '.data["id"]'\`
+image_status=\`oci image get --image-id=\${image_ocid} | jq -r '.data["lifecycle-state"]'\`
+while [ "\${image_status}" != "AVAILABLE" ]
```

Comment:
If something goes wrong with the image creation, this will be an infinite loop - same goes for the other loops in this script. Not blocking, but would be good put in a timeout. (Or is there some higher level timeout that could be applied to the entire ipxe script?)

Reply:
How long would you suggest?  The OS install time is indeterminate, so a good run may be interrupted and there is no way to deterministically detect completion.

Reply:
I don't have the necessarily data to put a number on this. I do know that image creation can take quite a while. 

Another option (that can be done in addition to or instead of) is to look for known failure states, such as if the lifecycle state goes to DELETED.

Again, nice to have some safety checks here, but should not be blocking.

Reply:
If there were the possibility of a safety check, I would have put it in.  Since the install process is completely blind, the only possible states of the system are AVAILABLE and PROVISIONING. I will put in a DELETED in the middle of the loop to allow for a manual destruction process.